### PR TITLE
Consent Reporting Client IP

### PR DESCRIPTION
**Pull request by pattisdr**
Closes #PROD-1412

### Description Of Changes

For legal purposes, sometimes a truncated version of the client ip is needed to prove what actions were taken after consent was saved.  Here we are running fides in such a way that the ip of the load balancer is being recorded, not the client ip.

### Code Changes

* [ ] Attempt to log first `x-forwarded-for` ip if it exists, otherwise, drop back to request.client.host

### Steps to Confirm

- Saved served notice history with x-forwarded-for header
```bash
curl -X 'PATCH' \
  'http://localhost:8080/api/v1/notices-served' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -H 'X-Forwarded-For: 123.123.123.123,1.1.1.1,2.2.2.2' \
  -d '{
  "browser_identity": {
    "fides_user_device_id": "208f6661-60ac-456a-8143-4954ff427b6c"
  },
  "tcf_purpose_consents": [1],
   "serving_component": "overlay"
}'
```

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`
* [ ] For API changes, the [Postman collection](https://github.com/ethyca/fides/blob/main/docs/fides/docs/development/postman/Fides.postman_collection.json) has been updated

### CODE REVIEW

File: src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py [hunk 1]
```diff
@@ -187,6 +188,26 @@ def get_request_origin_and_config() -> Tuple[Optional[str], Optional[str]]:
     )
 
 
+def get_ip_address(request: Request) -> Optional[str]:
+    """Get client ip, preferring x-forwarded-for if it exists, otherwise, dropping back to
+    request.client.host"""
+    x_forwarded_for = (
+        request.headers.get("x-forwarded-for") if request.headers else None
```
Comment (pattisdr) [inline src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py:195]: I am not sure if I need to separately allow `forwarded-allow-ips` to be configured as an env var to pass into uvicorn 

https://www.uvicorn.org/settings/#http

> --proxy-headers / --no-proxy-headers - Enable/Disable X-Forwarded-Proto, X-Forwarded-For, X-Forwarded-Port to populate remote address info. Defaults to enabled, but is restricted to only trusting connecting IPs in the forwarded-allow-ips configuration.

> --forwarded-allow-ips Comma separated list of IPs to trust with proxy headers. Defaults to the $FORWARDED_ALLOW_IPS environment variable if available, or '127.0.0.1'. A wildcard '*' means always trust.

File: src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py [hunk 2]
```diff
@@ -172,7 +172,8 @@ def get_request_origin_and_config() -> Tuple[Optional[str], Optional[str]]:
         return origin, experience_config_history_identifier
 
     request_headers = request.headers
-    ip_address: Optional[str] = request.client.host if request.client else None
+
+    ip_address: Optional[str] = get_ip_address(request)
```
Comment (pattisdr) [inline src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py:176]: Note Served Notice History endpoints are in  OSS but Privacy Preference History endpoints are in Plus.

They share this IP address code so if it works here, it should work over in Plus.

File: src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py [hunk 3]
```diff
@@ -187,6 +189,30 @@ def get_request_origin_and_config() -> Tuple[Optional[str], Optional[str]]:
...
 
 
+def get_ip_address(request: Request) -> Optional[str]:
+    """Get client ip, preferring x-forwarded-for if it exists, otherwise, dropping back to
+    request.client.host"""
+    x_forwarded_for = (
+        request.headers.get("x-forwarded-for") if request.headers else None
+    )
+
+    logger.info(f"Temporary log: X-Forwarded-For {x_forwarded_for}")
```
Comment (pattisdr) [inline src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py:199]: ❗ TODO remove before merge

File: src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py [hunk 4]
```diff
@@ -187,6 +189,30 @@ def get_request_origin_and_config() -> Tuple[Optional[str], Optional[str]]:
...
+    if x_forwarded_for:
+        try:
+            client_ip = x_forwarded_for.split(",")[0].strip()
+        except AttributeError:
+            pass
+
+    if not client_ip:
+        client_ip = request.client.host if request.client else None
+
+    logger.info(f"Temporary log: Client IP {client_ip}")
```
Comment (pattisdr) [inline src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py:211]: ❗ TODO remove before merge

File: src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py [hunk 5]
```diff
@@ -187,6 +189,30 @@ def get_request_origin_and_config() -> Tuple[Optional[str], Optional[str]]:
...
+    )
+
+    logger.info(f"Temporary log: X-Forwarded-For {x_forwarded_for}")
+
+    client_ip: Optional[str] = None
+    if x_forwarded_for:
+        try:
+            client_ip = x_forwarded_for.split(",")[0].strip()
+        except AttributeError:
+            pass
```
Comment (adamsachs) [inline src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py:206]: probably missing something obvious but what's the `AttibuteError` that can be thrown here? is it possible that the `x_forwarded_for` variable isn't a string?

File: src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py (hunk 1 above)
Comment (adamsachs) [reply src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py:196]: interesting, is that saying the _header_ is rejected if that's not configured, or the request in general? (this is probably obvious and i'm just missing context!)

seems unlikely we're configuring this already (i don't see any reference to the config keys anywhere in either fides/fidesplus/fidesinfra codebases)...but i'm just spitballing!

**Review by adamsachs — APPROVED** [review]
looks good to me! relatively straightforward in terms of app code changes, i think.

tentative approval but seems like we need to resolve your more deployment/config related question re: the uvicorn settings...

File: src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py (hunk 5 above)
Comment (pattisdr) [reply src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py:206]: I was just thinking if it didn't exist?

File: src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py (hunk 1 above)
Comment (pattisdr) [reply src/fides/api/api/v1/endpoints/privacy_preference_endpoints.py:196]: I can't tell from the docs! I am not sure if it's needed, I tested this alpha tag https://github.com/ethyca/fides/releases/tag/2.24.2a0 on a staging environment Robert set up, and it looks like X-Forwarded-For is getting passed through to the application just fine?

**Pull request merged**